### ISSUE TITLE
Bootstrapping does not know if nodes are ipv6 or TCP, tries invalid nodes

### ISSUE THREAD

**Issue by anthonybilinski** [qTox/qTox#5114]
The nodes on https://nodes.tox.chat have info like which TCP ports the bootstrap node supports and whether it's accessible over IPv4, IPv6, or both.
![image](https://user-images.githubusercontent.com/10469203/39495558-a457d0d4-4d4f-11e8-8ad8-d97478a31777.png)

Our DHT bootstrapping doesn't take this info into account. We simply have {Pub key, IP/domain, port} sets. This means that when qTox is in TCP only mode, we still end up trying to add bootstrap nodes as TCP relays even if they have no TCP ports, or we try to add them on their UDP port instead of their TCP port(s). We also try IPv6 nodes even when IPv6 is disabled. All of these invalid attemps slow down our bootstrap process since we only try 3 nodes every period of time.

qTox should contain more info on the bootstrap nodes, i.e. If it supports TCP and if so on which ports, and if it supports IPv4, IPv6, or both. This way we could only try to bootstrap nodes that fit our required settings. This would speed up bootstrapping for users in TCP-only mode with IPv6 disabled, i.e. all Tor users.

This would require some changes to `Core::bootstrapDht` to be settings-aware, and also changing the format of res/settings.ini to store additional info about the bootstrap nodes.

Related issue: #4084

**Comment by tox-user** [comment]
Currently connecting can take even a few minutes if someone is using Tor or can't use ipv6, because a lot of the nodes won't work and we try to connect to them anyway.

**Comment by sudden6** [comment]
When updating this, we should switch to using the JSON output under https://nodes.tox.chat/json directly.

**Comment by tox-user** [comment]
I don't think we have to use JSON as long as we have a script that will download and update the list for us for every release.

**Comment by sudden6** [comment]
@tox-user IMO we should use JSON, because that's the source format anyway. Makes it easier to update also, or when did someone last submit a PR to update bootstrap nodes? ;P

**Comment by tox-user** [comment]
There was this: https://github.com/qtox/qtox/issues/4690. I don't know of any newer ones.

**Comment by anthonybilinski** [comment]
@tox-user found a relevant doc on client bootstrapping: https://github.com/Tox-Clients/Tox_Client_Guidelines/blob/master/Required/Bootstrapping.md.
Interestingly,
>When loading from a save, you do not need to bootstrap unless ~10 seconds after running the first tox_iterate() toxcore still doesn't report that it is connected

which I don't think we follow.

**Comment by sudden6** [comment]
Nope, definitely not, is this document still up to date with the current toktok/c-toxcore?

**Comment by anthonybilinski** [comment]
May well not be, hasn't been updated since 2015. Might be worth just benchmarking both methods or looking at toxcore source / other clients.

**Comment by anthonybilinski** [comment]
Regarding the last post, I talked to some toxcore guys about how to bootstrap recently.
* If the profile is brand new and there are no saved nodes, the client should add many bootstrap nodes.
* If the profile has saved nodes, but hasn't been opened for a while so these nodes are probably stale, bootstrap nodes should be added quickly.
* If the profile was recently opened, bootstrapping may slow down connection time.

The client has no way to tell which of these cases they have. There was talk about adding a callback to toxcore to request bootstrap nodes, so that based on network health it could be called rapidly or not at all.

Our current behaviour doesn't seem unreasonable based on the API available.

**Comment by anthonybilinski** [comment]
After https://github.com/qTox/qTox/pull/6126 Core now has access to which nodes support TCP and UDP, and which are IPv4 and IPv6. The bootstrapping logic hasn't yet been updated to e.g. add two TCP relays per period if in TCP mode, but it wouldn't be much work form here.
